cpp/src/arrow/result_test.cc

```diff
@@ -539,6 +539,45 @@ TEST(ResultTest, StatusReturnAdapterMoveAndConvertValue) {
   EXPECT_EQ(*implicitly_convertible.move_only.data, kIntElement);
 }
 
+// Verify that a Result<T> can be queried for a stored value or an alternative.
+TEST(ResultTest, ValueOrAlternative) {
+  EXPECT_EQ(Result<MoveOnlyDataType>(MoveOnlyDataType{kIntElement})
+                .ValueOr(MoveOnlyDataType{0})
+                .data[0],
+            kIntElement);
+
+  EXPECT_EQ(
+      Result<MoveOnlyDataType>(Status::Invalid("")).ValueOr(MoveOnlyDataType{0}).data[0],
+      0);
+}
+
+TEST(ResultTest, MapFunctionToMoveOnlyDataType) {
+  Result<MoveOnlyDataType> result(MoveOnlyDataType{kIntElement});
+
+  static auto error = Status::Invalid("some error message");
+
+  auto const_mapped =
+      result.Map([](const MoveOnlyDataType& m) -> Result<int> { return *m.data; });
+  EXPECT_TRUE(const_mapped.ok());
+  EXPECT_EQ(const_mapped.ValueOrDie(), kIntElement);
+
+  auto const_error =
+      result.Map([](const MoveOnlyDataType& m) -> Result<int> { return error; });
```

Comment:
How does `return error` work if you're not capturing anything?

Reply:
`error` is static so it doesn't require capture

Reply:
Ahah, a C++ oddity I didn't know about :-)